core/src/main/scala/chisel3/Printable.scala

```diff
@@ -45,6 +45,8 @@ import java.util.{MissingFormatArgumentException, UnknownFormatConversionExcepti
 // TODO Should we provide more functions like map and mkPrintable?
 sealed abstract class Printable {
 
+  val context = internal.Builder.currentModule
```

Comment:
The context/lexical scope we need to check is not a property of the `Printable` object, rather it's a property of the `Data` boxed by various `Printables` (eg. `Decimal`). Thus `Printable` itself should not be capturing the Module in which it was created. Furthermore, lexical scope also needs to take `when` scoping into account, see this example Scastie (https://scastie.scala-lang.org/yJB6k07dSm6rTian91M1rA)

As a straw man, I think what you want to do is have a [package private] method (maybe `def checkLexicalScope(moduleContext: RawModule, whenContext: Option[WhenContext]): Unit`) that recurses through the `Printable` structure and checks the lexical scope of each `Data`, reporting an error if the scope is violated.